algorithmic_efficiency/logging_utils.py

```diff
@@ -0,0 +1,492 @@
+  if gpus:
+    gpu_count = len(gpus)
+    measurements['gpu.count'] = gpu_count
+    avg_gpu_load = 0
+    avg_gpu_memory_util = 0
+    avg_gpu_memory_total = 0
+    avg_gpu_memory_used = 0
+    avg_gpu_memory_free = 0
+    avg_gpu_temperature = 0
+    for gpu in gpus:
+      idx = gpu.id
+      measurements[f'gpu.{idx}.compute.util'] = gpu.load
+      measurements[f'gpu.{idx}.mem.util'] = gpu.memoryUtil
+      measurements[f'gpu.{idx}.mem.total'] = gpu.memoryTotal
+      measurements[f'gpu.{idx}.mem.used'] = gpu.memoryUsed
+      measurements[f'gpu.{idx}.mem.free'] = gpu.memoryFree
+      measurements[f'gpu.{idx}.temp.current'] = gpu.temperature
+      # Note: GPU wattage was not available from gputil as of writing
+      avg_gpu_load += gpu.load
+      avg_gpu_memory_util += gpu.memoryUtil
+      avg_gpu_memory_total += gpu.memoryTotal
+      avg_gpu_memory_used += gpu.memoryUsed
+      avg_gpu_memory_free += gpu.memoryFree
+      avg_gpu_temperature += gpu.temperature
+    measurements['gpu.avg.compute.util'] = avg_gpu_load / gpu_count
+    measurements['gpu.avg.mem.util'] = avg_gpu_memory_util / gpu_count
+    measurements['gpu.avg.mem.total'] = avg_gpu_memory_total / gpu_count
+    measurements['gpu.avg.mem.used'] = avg_gpu_memory_used / gpu_count
+    measurements['gpu.avg.mem.free'] = avg_gpu_memory_free / gpu_count
+    measurements['gpu.avg.temp.current'] = avg_gpu_temperature / gpu_count
+
+  return measurements
+
+
+def _get_git_commit_hash() -> str:
+  return subprocess.check_output(['git', 'rev-parse',
+                                  'HEAD']).decode('ascii').strip()
+
+
+def _get_git_branch() -> str:
+  return subprocess.check_output(['git', 'branch',
+                                  '--show-current']).decode('ascii').strip()
+
+
+def _get_cpu_model_name() -> str:
+  output = subprocess.check_output(['lscpu']).decode('ascii').strip()
+  return re.findall(r"(?=Model name:\s{1,}).*",
+                    output)[0].split('Model name:')[1].strip()
+
+
+def _get_os_package_list() -> str:
+  return subprocess.check_output(['dpkg', '-l']).decode('ascii').strip()
+
+
+def _get_pip_package_list() -> str:
+  return subprocess.check_output(['pip', 'freeze']).decode('ascii').strip()
+
+
+def _is_primitive_type(item: Any) -> bool:
+  primitive = (float, int, str, bool)
+  return isinstance(item, primitive)
+
+
+def _get_extra_metadata_as_dict(extra_metadata_string_list: list) -> dict:
+  """Parse the extra_metadata CLI argument from string into dict.
+
+  For example this program was executed with --record_extra_metadata="key=value"
+  then {'extra.key':'value'} is returned.
+  """
+  metadata = {}
+  if not extra_metadata_string_list:
+    return metadata
+  try:
+    for item in extra_metadata_string_list:
+      key, value = item.split("=")
+      metadata['extra.' + key] = value
+  except:
+    logging.error(
+        'Failed to parse extra_metadata CLI arguments. Please check your '
+        'command.')
+    raise
```

Comment:
it could be more helpful to move the try/except into the for loop, and then if you can't parse the error then raise an exception and include the item that was unable to be parsed in the exception message.

Also instead of logging and a naked `raise`, I would do something like `raise ValueError('Failed to par...')`

Reply:
Fixed.